### PR TITLE
feat(widget): widgets emit widget.effects

### PR DESCRIPTION
**Pull request by EdAyers**
save_widget now has the type `pos -> component tactic_state
widget.effects -> tactic unit`.
This means that now the widget-writer can cause widget.effects
to issue a stream of commands to vscode to run side effects. Things like
highlighting a particular region of code or revealing a location in the
editor.

The next task is to add 'go to definition' for the default tactic
widget.

It should be backwards compatible with older versions, both for vscode and lean.

### CODE REVIEW

File: library/init/meta/widget/basic.lean [hunk 1]
```diff
@@ -279,19 +279,30 @@ meta def with_cn : string → html α → html α
 meta def with_key {β} [has_to_string β] : β → html α → html α
 | s h := with_attr (key s) h
 
+/-- An effect is an action at the root of the widget component hierarchy
+and can give instructions to the editor to perform some task. -/
+meta inductive effect : Type
+| insert_text (text : string)
+| reveal_position (file_name : string) (p : pos)
```
Comment (gebner) [inline library/init/meta/widget/basic.lean:286]: Is empty string = current file?  We cannot get the file name of the current file in meta code.

File: library/init/meta/widget/basic.lean [hunk 2]
```diff
@@ -279,19 +279,30 @@ meta def with_cn : string → html α → html α
...
+| highlight_position (file_name : string) (p : pos)
+| clear_highlighting
+| custom (key : string) (value : string)
```
Comment (gebner) [inline library/init/meta/widget/basic.lean:289]: What do you think the `custom` effect will be used for?  You need to modify the vscode extension anyhow for a new effect type.

Comment (EdAyers) [reply library/init/meta/widget/basic.lean:289]: Yeah it just means that you can test out new effects without having to have a custom build of lean was the idea.

File: src/frontends/lean/widget.cpp [hunk 3]
```diff
@@ -315,6 +341,56 @@ std::vector<vdom> render_html_list(vm_obj const & htmls, std::vector<component_i
...
+    for (auto e : effects) {
+        switch (cidx(e)) {
+            case effect_idx::insert_text: {
+                result.push_back({
+                    {"kind", "insert_text"},
+                    {"text", to_string(cfield(e, 0))}
+                });
+                break;
+            } case effect_idx::reveal_position: {
+                auto pos = to_pos_info(cfield(e, 1));
```
Comment (EdAyers) [inline src/frontends/lean/widget.cpp:365]: I get a vm_check failed here because `cfield(e,1)` is not a vm_composite for the `widget_effects` test. :/

**Pull request closed without merging**